onecodex/scripts/export_functional_metric.py

```diff
@@ -0,0 +1,301 @@
+
+            value = item["value"]
+            row.append(f"{value:.3f}" if value is not None else "")
+            self.csv_writer.writerow(row)
+
+    def produce_output(self):
+        self.fd.close()
+
+
+class WideFunctionalResultExporter(BaseFunctionalResultExporter):
+    def __init__(
+        self,
+        *,
+        out_path: str | None = None,
+        file_obj: TextIO | None = None,
+        taxa_stratified: bool,
+        metric: str,
+    ):
+        if out_path is None and file_obj is None:
+            raise OneCodexException("Provide one of out_path or file_obj")
+        elif out_path is not None and file_obj is not None:
+            raise OneCodexException("Provide only one of out_path or file_obj")
+
+        self.fd = open(out_path, "w", newline="") if out_path is not None else file_obj
+        self.csv_writer = csv.writer(self.fd)
+        self.taxa_stratified = taxa_stratified
+        self.metric = FunctionalAnnotationsMetric(metric)
+
+        self.headers = (
+            [
+                "Function ID",
+                "Function Name",
+                "Taxon ID",
+                "Taxon Name",
+            ]
+            if taxa_stratified
+            else [
+                "Function ID",
+                "Function Name",
+            ]
+        )
+
+        self.functional_run_positions = {}
+        self.function_names = {}
+        self.temp_files = {}
+        self.temp_dir = tempfile.mkdtemp(prefix="func_exp_")
+
+    def consume_results(self, sample_id: str, sample_name: str, analysis_id: str, results: dict):
+        if analysis_id not in self.functional_run_positions:
+            self.functional_run_positions[analysis_id] = len(self.headers)
+            self.headers.append(f"{sample_name} (run id {analysis_id})")
+
+        for item in results.get("table", []):
+            function_id = item["id"]
+            if function_id not in self.function_names:
+                self.function_names[function_id] = item["name"]
+
+            # Split data up either by function_id or by (function_id, taxon_name, taxon_id)
+            if self.taxa_stratified:
+                taxon_id = str(item.get("taxon_id") or "")
+                taxon_name = parse_functional_taxon_name(item.get("taxon_name") or "")
+                key = (function_id, taxon_name, taxon_id)
+            else:
+                key = function_id
+
+            if key not in self.temp_files:
+                tmp_path = os.path.join(self.temp_dir, f"{hash_to_hex(key)}.csv")
+                self.temp_files[key] = tmp_path
+            else:
+                tmp_path = self.temp_files[key]
+
+            value = item["value"]
+
+            # Save data to a temp file in long format
+            with open(tmp_path, "a", newline="") as tmp_file:
+                tmp_writer = csv.writer(tmp_file)
+                tmp_writer.writerow([analysis_id, f"{value:.3f}" if value is not None else ""])
```

Comment:
I think we should place 0s in cells where there are no values. cc @audy or @treynr for alternative suggestions.